packages/labs/observers/src/intersection_controller.ts

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+import {ReactiveControllerHost} from '@lit/reactive-element/reactive-controller.js';
+
+export type IntersectionValueCallback = (
+  ...args: Parameters<IntersectionObserverCallback>
+) => unknown;
+
+export interface IntersectionControllerConfig {
+  config?: IntersectionObserverInit;
+  target?: Element;
+  callback?: IntersectionValueCallback;
+  skipInitial?: boolean;
+}
+
+export class IntersectionController {
+  private _host: ReactiveControllerHost;
+  private _target?: Element | null;
+  private _observer: IntersectionObserver;
+  private _skipInitial = false;
+  private _unobservedUpdate = false;
+  value?: unknown;
+  callback: IntersectionValueCallback = () => true;
+  constructor(
+    host: ReactiveControllerHost,
+    {target, config, callback, skipInitial}: IntersectionControllerConfig
+  ) {
+    this._host = host;
+    this._target = target;
+    if (skipInitial !== undefined) {
+      this._skipInitial = skipInitial;
+    }
+    if (callback !== undefined) {
+      this.callback = callback;
+    }
+    this._observer = new IntersectionObserver(
+      (entries: IntersectionObserverEntry[]) => {
+        const unobservedUpdate = this._unobservedUpdate;
+        this._unobservedUpdate = false;
+        if (this._skipInitial && unobservedUpdate) {
+          return;
+        }
+        this.handleChanges(entries);
+        this._host.requestUpdate();
+      },
+      config
+    );
+    this._host.addController(this);
+  }
+
+  protected handleChanges(entries: IntersectionObserverEntry[]) {
+    this.value = this.callback(entries, this._observer);
+  }
+
+  hostConnected() {
+    if (this._target) {
+      this.observe(this._target);
+    }
+  }
+
+  hostDisconnected() {
+    this.disconnect();
+  }
+
+  async hostUpdated() {
+    // Eagerly deliver any changes that happened during update.
+    const pendingRecords = this._observer.takeRecords();
+    if (pendingRecords.length) {
+      this.handleChanges(pendingRecords);
+    }
+  }
+
+  /**
+   * Observe the target element.
+   * @param target Element to observe
+   */
+  observe(target: Element) {
```

Comment:
Do we want to reserve `observe()` for a future directive, ie:

```ts
html`<div ${this._intersectionController.observe()}></div>
```

Reply:
I was thinking we could have 1 generic `observeWith` directive for this package that does this:

`html`<div ${observeWith(this._intersectionController)}></div>`

That directive would use the controller's `observe` method. That way the controller's don't rely on lit-html specifically. Wdyt?

Reply:
hmm... that seems slightly less ergonomic for the Lit users (no code completion tells them the API is there) , but I get the desire to decouple from lit-html

Reply:
Added to https://github.com/lit/lit/issues/2350.